content/articles/haskell-websocket-tsung-benchmark.md

```diff
@@ -0,0 +1,279 @@
+Title: A lot of websockets in Haskell
+Date: 2015-11-05
+Short_summary: I implement a very simple Haskell websocket chat server and benchmark it with tsung.
+Category: Dev
+Authors: Tom
+
+*Reading time: ~20 minutes.*
+
+I really enjoyed reading how the phoenix-framework people managed to
+get to
+[two million active websocket connections](http://www.phoenixframework.org/v1.0.0/blog/the-road-to-2-million-websocket-connections).
+
+I've heard some very smart people say that Haskell has an amazing
+runtime with very cheap threads. I have no reason to disbelieve them
+but we thought it'd be fun to see how Haskell fares in a comparable
+setup.
+
+<!-- PELICAN_END_SUMMARY -->
+
+## Test servers
+
+Unlike the Phoenix people we don't have Rackspace sponsorship so we had
+to resort to the common man's cheap machines: EC2 spot instances. We
+bid $0.10 on two `m4.xlarge` machines with 16G of RAM and 4 cores
+which usually cost around $0.05 per hour in eu-west.
+
+[As usual](https://blog.wearewizards.io/how-to-use-nixops-in-a-team)
+we're using Nix to deploy tsung and a very simple Haskell chat program
+that just broadcasts incoming messages to everyone.
+
+> **tsung** is a tcp/web load generator written in Erlang configured
+> through a XML domain-specific language
+> ([website](tsung.erlang-projects.org)).
+
+The core handler of our chat program looks like this (full source [here](https://github.com/WeAreWizards/haskell-websockets-tsung-benchmark/blob/master/code/src/Main.hs)):
+
+```haskell
+handleWS :: InChan ByteString -> PendingConnection -> IO ()
+handleWS bcast pending = do
+    localChan <- dupChan bcast
+    connection <- acceptRequest pending
+
+    -- Yes, we're leaking resources here because this forkIO
+    -- doesn't terminate when the socket closes :)
+    forkIO $ forever $ do
+        message <- readChan localChan
+        sendTextData connection message
+
+    -- loop forever
+    let loop = do
+            Text message <- receiveDataMessage connection
+            writeChan bcast message
+            loop
+    loop
+```
+
+To run the ec2 machines we're using
+[nixops](https://nixos.org/nixops/) which also does the spot-price
+bidding for us:
+
+```
+nixops create '<nix/test-setup.nix>'
+nixops deploy
+```
+
+(See
+[here](https://github.com/WeAreWizards/haskell-websockets-tsung-benchmark/blob/master/nix/test-setup.nix)
+for the full configuration including kernel tuning).
+
+## tsung setup
+
+Unfortunately I could not get the distributed tsung going: The
+distributed testing uses an Erlang function called `slave:start` which
+connects through SSH and spawns Erlang on the remote host. This failed
+for reasons I didn't have time to debug.
+
+But without the distributed testing there's a problem: A single server
```

Comment:
I think a full sentence after a colon should start with a capital letter but both styles are allowed (e.g. [1])

[1]
http://www.bristol.ac.uk/arts/exercises/grammar/grammar_tutorial/page_04.htm